R/readConfig.R

```diff
@@ -5,172 +5,243 @@
+#'   BRICK-internal config folder is used.
+#' @param readDirect logical, specify whether \code{config} is a valid path to a
+#'   config file that should be read directly.
 #' @returns named list with run config
 #'
 #' @author Robin Hasse
-#'
-#' @importFrom yaml read_yaml
 
-readConfig <- function(config = NULL, basisOf = NULL, readDirect = FALSE) {
+readConfig <- function(config = NULL, configFolder = NULL, readDirect = FALSE) {
 
-  # check input ----------------------------------------------------------------
-
-  # list all config files this config is based on
-  readCfg <- function(file) {
-    cfg <- read_yaml(file)
-    if (!"title" %in% names(cfg)) {
-      stop("There is no title given in the config file: ", file)
-    }
-    attr(cfg, "file") <- file
-    attr(cfg, "files") <- file
-    return(cfg)
+  # check config files that are based on the current to avoid circle dependency
+  basisOf <- attr(config, "basisOf")
+  if (any(duplicated(basisOf))) {
+    stop("There is a circle dependency in how configs are based on another: ",
+         paste(basisOf, collapse = " -> "))
   }
 
-  # Directly read the file in the given config file path - used after the final config has already been compiled
+  # Directly read the file without considering basis configs
   if (readDirect) {
     if (!file.exists(config)) {
-      stop("The config", config, "that was called to be read directly does not exist. ",
-           "If this is a restart run, your folder to restart in likely misses the config file.")
+      stop("The config", config, "that you want read directly does not exist. ",
+           "If this is a restart run, your run folder likely misses the config file.")
+    }
+    if (!is.null(configFolder)) {
+      warning("'configFolder' is ignored as the config is read directly from here: ",
+              config)
     }
-    return(readCfg(config))
+    return(.readCfg(config))
   }
 
-  configFolder <- brick.file("config")
+  # use default in config folder if available, otherwise internal default
+  internalConfigFolder <- brick.file("config")
+  if (is.null(configFolder)) {
+    configFolder <- internalConfigFolder
+  } else {
+    if (!dir.exists(configFolder)) {
+      stop("Directory of 'configFolder' doesn't exist: ", configFolder)
+    }
+  }
   defaultCfgPath <- file.path(configFolder, "default.yaml")
-
   if (!file.exists(defaultCfgPath)) {
-    stop("Default config ", defaultCfgPath, " does not exist.")
+    defaultCfgPath <- file.path(internalConfigFolder, "default.yaml")
   }
 
   # use default.yaml by default
   if (is.null(config)) {
     if (is.null(basisOf)) {
+      config <- defaultCfgPath
       message("Using default config: ", config)
     }
   }
 
   # find file to given config
+  customCfgPath <- .findCfg(config, configFolder, isFinalCfg = is.null(basisOf))
+
+  # read yaml
+  customCfg <- .readCfg(customCfgPath)
+  if (file.path(customCfgPath) == file.path(defaultCfgPath)) {
+    return(.readCfg(defaultCfgPath))
+  }
```

Comment:
Maybe you could add a comment here that this is the code snippet that ends the recursion? As it ensures that we have arrived at the default.yaml? (If I got it correctly...) Although I just realized that you didn't actually change this, but still an additional comment would improve readability of the code.

Reply:
Good point. I did that.